### PR TITLE
[SYCL] Remove exception ctors referencing std::string from ABI boundary

### PR DESCRIPTION
**Pull request by bso-intel**
This PR ABI-neutralize some exception ctors that require context objects.
However, #include <sycl/context.hpp> caused a cyclic dependency issue.
So, moving them into context.hpp solved the issue.

### CODE REVIEW

File: sycl/test/abi/sycl_symbols_linux.dump [hunk 1]
```diff
@@ -4093,14 +4093,14 @@ _ZNK4sycl3_V16device9getNativeEv
...
-_ZNK4sycl3_V16kernel16get_backend_infoINS0_4info6device7versionEEENS0_6detail20is_backend_info_descIT_E11return_typeEv
-_ZNK4sycl3_V16kernel16get_backend_infoINS0_4info8platform7versionEEENS0_6detail20is_backend_info_descIT_E11return_typeEv
 _ZNK4sycl3_V16kernel13get_info_implINS0_4info6kernel10attributesEEENS0_6detail11ABINeutralTINS6_19is_kernel_info_descIT_E11return_typeEE4typeEv
 _ZNK4sycl3_V16kernel13get_info_implINS0_4info6kernel13function_nameEEENS0_6detail11ABINeutralTINS6_19is_kernel_info_descIT_E11return_typeEE4typeEv
 _ZNK4sycl3_V16kernel13get_info_implINS0_4info6kernel15reference_countEEENS0_6detail11ABINeutralTINS6_19is_kernel_info_descIT_E11return_typeEE4typeEv
 _ZNK4sycl3_V16kernel13get_info_implINS0_4info6kernel7contextEEENS0_6detail11ABINeutralTINS6_19is_kernel_info_descIT_E11return_typeEE4typeEv
 _ZNK4sycl3_V16kernel13get_info_implINS0_4info6kernel8num_argsEEENS0_6detail11ABINeutralTINS6_19is_kernel_info_descIT_E11return_typeEE4typeEv
+_ZNK4sycl3_V16kernel16get_backend_infoINS0_4info6device15backend_versionEEENS0_6detail20is_backend_info_descIT_E11return_typeEv
```
Comment (aelovikov-intel) [inline sycl/test/abi/sycl_symbols_linux.dump:4101]: can drop this change.

**Review by aelovikov-intel — APPROVED** [review]

File: sycl/include/sycl/context.hpp [hunk 2]
```diff
@@ -260,6 +260,35 @@ class __SYCL_EXPORT context : public detail::OwnerLessBase<context> {
   friend T detail::createSyclObjFromImpl(decltype(T::impl) ImplObj);
 };
 
+#ifdef __INTEL_PREVIEW_BREAKING_CHANGES
+// context.hpp depends on exception.hpp but we can't define these ctors in
+// exception.hpp while context is still an incomplete type.
```
Comment (sergey-semenov) [inline sycl/include/sycl/context.hpp:265]: Can we leave a similar comment pointing to context.hpp in exception.hpp?

Comment (bso-intel) [reply sycl/include/sycl/context.hpp:265]: Done. Thanks.

File: sycl/test/abi/sycl_symbols_linux.dump (hunk 1 above)
Comment (bso-intel) [reply sycl/test/abi/sycl_symbols_linux.dump]: Done. Thanks.

**Review by sergey-semenov — APPROVED** [review]

**Comment by bso-intel** [comment]
@intel/llvm-gatekeepers 
Please merge this PR.

**Comment by bso-intel** [comment]
> Since the ABI-break is behind the preview flag, could you please remove the tag and "[ABI-Break]" part of the title?

I was asked to add the tag since this is an ABI-breaking changes.
Although it is enabled under the flag -fpreview-breaking-changes. It is truly an ABI-breaking change.

**Comment by bso-intel** [comment]
I removed the tag [ABI-Break] and the label.
Please merge this PR.
Thanks.

**Comment by steffenlarsen** [comment]
> > Since the ABI-break is behind the preview flag, could you please remove the tag and "[ABI-Break]" part of the title?
> 
> I was asked to add the tag since this is an ABI-breaking change. Although it is enabled under the flag -fpreview-breaking-changes. It is truly an ABI-breaking change.

Though you are right that it breaks ABI behind the preview flag, we don't normally consider those ABI breaks as that is what the preview flag is intended for. I would prefer we save the label and title tag for cases where we need to be aware of it breaking ABI, e.g. for stalling patches until an ABI-break window. @aelovikov-intel & @sergey-semenov - Thoughts?

**Comment by sergey-semenov** [comment]
> > > Since the ABI-break is behind the preview flag, could you please remove the tag and "[ABI-Break]" part of the title?
> > 
> > 
> > I was asked to add the tag since this is an ABI-breaking change. Although it is enabled under the flag -fpreview-breaking-changes. It is truly an ABI-breaking change.
> 
> Though you are right that it breaks ABI behind the preview flag, we don't normally consider those ABI breaks as that is what the preview flag is intended for. I would prefer we save the label and title tag for cases where we need to be aware of it breaking ABI, e.g. for stalling patches until an ABI-break window. @aelovikov-intel & @sergey-semenov - Thoughts?

Agreed, I don't think it makes sense to use the tag for changes like this one.

**Pull request merged**